From e25edd45bcbd229cc1de1a26c787bcb99db1e1ec Mon Sep 17 00:00:00 2001
From: Cajus Pollmeier <pollmeier@gonicus.de>
Date: Thu, 22 Oct 2026 14:04:47 +0200
Subject: [PATCH] Quick fixes for

 * timezone / UTC parsing
 * disable global DEBUG (as it prints credentials to the log)
 * fix crash on empty replies
---
 CMakeLists.txt       |  1 -
 qwebdav.cpp          |  5 +++++
 qwebdavdirparser.cpp | 16 ++++++++++++++--
 3 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/qwebdav.cpp b/qwebdav.cpp
index 091500d..34d5004 100644
--- a/qwebdav.cpp
+++ b/qwebdav.cpp
@@ -237,6 +237,10 @@ void QWebdav::provideAuthenication(QNetworkReply *reply, QAuthenticator *authent
     }
 #endif
 
+    if (reply == nullptr) {
+        return;
+    }
+
     if (reply == m_authenticator_lastReply) {
         reply->abort();
         emit errorChanged("WebDAV server requires authentication. Check WebDAV share settings!");
@@ -699,6 +703,7 @@ QNetworkRequest QWebdav::buildRequest() {
         QByteArray data = concatenated.toLocal8Bit().toBase64();
         QString headerData = "Basic " + data;
         req.setRawHeader("Authorization", headerData.toLocal8Bit());
+        req.setAttribute(QNetworkRequest::RedirectPolicyAttribute, QNetworkRequest::SameOriginRedirectPolicy);
     }
     return req;
 }
diff --git a/qwebdavdirparser.cpp b/qwebdavdirparser.cpp
index 15cfc99..1f7f5e8 100644
--- a/qwebdavdirparser.cpp
+++ b/qwebdavdirparser.cpp
@@ -309,8 +309,13 @@ void QWebdavDirParser::parseMultiResponse(const QByteArray &data)
         if (m_abort)
             return;
 
+        const auto itemCount = m_dirList.size();
         parseResponse(thisResponse);
 
+        // Noting changed
+        if (m_dirList.size() == itemCount)
+            continue;
+
         // container without slash at the end is a wrong answer
         // remove this item from the list
         if ((!m_includeRequestedURI) && m_dirList.last().isDir() && !responseName.endsWith("/")) {
@@ -521,20 +526,27 @@ QDateTime QWebdavDirParser::parseDateTime(const QString &input, const QString &t
         return datetime;
 
     datetime = usLocal.toDateTime(input.left(25), "ddd, dd MMM yyyy hh:mm:ss");
+    if (datetime.isValid()) {
+        datetime.setTimeZone(QTimeZone::utc());
+        return datetime;
+    }
+    datetime = QDateTime::fromString(input, Qt::ISODate);
     if (datetime.isValid())
         return datetime;
     datetime = usLocal.toDateTime(input.left(19), "yyyy-MM-ddThh:mm:ss");
     if (datetime.isValid())
         return datetime;
     datetime = usLocal.toDateTime(input.mid(5, 20) , "d MMM yyyy hh:mm:ss");
-    if (datetime.isValid())
+    if (datetime.isValid()) {
+        datetime.setTimeZone(QTimeZone::utc());
         return datetime;
+    }
     QDate date;
     QTime time;
 
     date = usLocal.toDate(input.mid(5, 11) , "d MMM yyyy");
     time = usLocal.toTime(input.mid(17, 8) , "hh:mm:ss");
-    datetime = QDateTime(date, time);
+    datetime = QDateTime(date, time, QTimeZone::utc());
 
 #ifdef DEBUG_WEBDAV
     if(!datetime.isValid())
-- 
2.56.0

